client/network/src/service.rs

```diff
@@ -1133,7 +1133,7 @@ struct Metrics {
 	kbuckets_num_nodes: GaugeVec<U64>,
 	listeners_local_addresses: Gauge<U64>,
 	listeners_errors_total: Counter<U64>,
-	network_per_sec_bytes: GaugeVec<U64>,
+	network_bytes_total: GaugeVec<U64>,
```

Comment:
```suggestion
	network_bytes_total: CounterVec<U64>,
```

`Gauge` is for metrics that go up and down, and `Counter` is for metrics that only ever increase over time except for restarts.

Reply:
Maybe I'm mistaken, but from the `prometheus` crate docs, counters have essentially `inc()` and `inc_by()`. Gauges can be `set`, which is what we need to do for these totals, i.e. these are counter values from existing counters (the running totals) reported as a gauge over time, so this is a monotonic gauge. I'm happy to be corrected.

Reply:
I see. The problem is that this is not only a change in the programmatic API of the `prometheus` library.
Whether a metric is a counter or a gauge is then reported to the Prometheus service, and then to Grafana.

Since we know that the value can only every increase, we could use [`Counter::get`](https://docs.rs/prometheus/0.9.0/prometheus/core/struct.GenericCounter.html#method.get) to get the current value, then call `inc_by`.
Doing this would be racy if there were multiple clones of that `Counter`, but since we only ever access that `Counter` from a single location, we know that nothing will modify the counter between the call to `get` and the call to `inc_by`.

@mxinden Any opinion?

Reply:
The problem of _updating a counter from an existing counter_ is often discussed in the Prometheus ecosystem. In general we suggest to implement a custom collector, see [robustperception](https://www.robustperception.io/setting-a-prometheus-counter) for general information, https://github.com/tikv/rust-prometheus/pull/303 specifically for the Rust client and my comment at the very end https://github.com/tikv/rust-prometheus/pull/303#issuecomment-647395752.

The reason why `Counter` does not implement `set` is, that it allows newcomers to easily shoot themselfes in the foot and thus it was not deemed worth the additional ergonomic for this advanced use case.

I see 3 ways moving forward:

- Keep it as a `Gauge`. This would expose the metric with the `TYPE gauge` header. As @tomaka mentioned above this is ingested by Prometheus and forwarded to Grafana, but only used for query suggestions, thus not a big issue.

- Use a `Counter`. As mentioned by @tomaka this is potentially racy if it ever gets set concurrently.

- Use a custom `impl Collector`. Quite verbose.

I would be fine with using a `Gauge` here for now. I would just ask to add a comment above stating that it is monotonic. In the future we can look into a custom `Collector`.

Reply:
@mxinden Thanks for your input.

> I would be fine with using a Gauge here for now. I would just ask to add a comment above stating that it is monotonic. In the future we can look into a custom Collector.

I added a comment to the code. I don't have a strong preference in any case.

Out of curiosity, if gauges "must go up and down" and counters "must only increase", what do you use for a monotonically decreasing metric? I'm not sure I understand the significance of the discussions of counter vs gauge. It would almost seem more sensible to directly offer and use [`Atomic`](https://docs.rs/prometheus/0.9.0/prometheus/core/trait.Atomic.html) for both use-cases, designating it as a "gauge" or "counter" based on the semantic meaning of the metric. I guess I just don't find the distinction between gauge and counter very useful, technically, especially if the counter can "jump", both forward and back to 0.

Reply:
> what do you use for a monotonically decreasing metric

So far I have never come across a use case for a monotonically decreasing metric. Can you come up with something that can not also be modeled by inverting a monotonically increasing counter?

> I'm not sure I understand the significance of the discussions of counter vs gauge. 

The benefit of differentiating the two comes into play only at query (aggregation time). E.g. one can apply the [`rate`](https://prometheus.io/docs/prometheus/latest/querying/functions/#rate) function to a counter knowing that the underlying data never breaks monotonicity. I think the best summary is given by [robust perception here](https://www.robustperception.io/how-does-a-prometheus-counter-work).

Let me know if the above makes sense @romanb. More than happy to put more effort into a detailed explanation.

Reply:
> So far I have never come across a use case for a monotonically decreasing metric. [..]

A metric for a bounded resource and its exhaustion over time? I don't know. The restriction not to be able to do so just seems a bit arbitrary.

> The benefit of differentiating the two comes into play only at query (aggregation time). E.g. one can apply the rate function to a counter knowing that the underlying data never breaks monotonicity. I think the best summary is given by robust perception here.

Ok, so there are special query functions with specific characteristics for gauges and counters, like `rate()` (counter) vs `deriv()` for gauges, both essentially yielding the derivative but the former makes special use of counter properties to better handle things like resets on restarts. Thanks for the pointers @mxinden.